File: py/candlestick_patterns/patterns/takuri.py
"""Takuri pattern (1-candle)."""
from __future__ import annotations

from ..core.primitives import real_body, upper_shadow, lower_shadow
from ...fuzzy import t_product_all


def takuri(self) -> float:
    """Takuri (Dragonfly Doji with very long lower shadow): a one-candle pattern.

    A doji body with a very short upper shadow and a very long lower shadow.

    Returns:
        Continuous float in [0, 100].  Higher = stronger signal.
    """
    if not self._enough(1, self._doji_body, self._very_short_shadow,
                        self._very_long_shadow):
        return 0.0

    o, h, l, c = self._bar(1)

    mu_doji = self._mu_less(real_body(o, c), self._doji_body, 1)
    mu_short_us = self._mu_less(upper_shadow(o, h, c), self._very_short_shadow, 1)
    mu_long_ls = self._mu_greater(lower_shadow(o, l, c), self._very_long_shadow, 1)

    confidence = t_product_all(mu_doji, mu_short_us, mu_long_ls)
    return confidence * 100.0

File: ts/candlestick-patterns/patterns/takuri.ts
/** Takuri pattern (1-candle). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, upperShadow, lowerShadow } from '../core/primitives.ts';

/**
 * Takuri (Dragonfly Doji with very long lower shadow): a one-candle pattern.
 *
 * A doji body with a very short upper shadow and a very long lower shadow.
 *
 * Returns:
 *     Continuous float in [0, 100].  Higher = stronger signal.
 */
export function takuri(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(1, cp.dojiBody, cp.veryShortShadow, cp.veryLongShadow)) return 0.0;

    const b = cp.bar(1);

    const muDoji = cp.muLessCS(realBody(b.o, b.c), cp.dojiBody, 1);
    const muShortUS = cp.muLessCS(upperShadow(b.o, b.h, b.c), cp.veryShortShadow, 1);
    const muLongLS = cp.muGreaterCS(lowerShadow(b.o, b.l, b.c), cp.veryLongShadow, 1);

    return tProductAll(muDoji, muShortUS, muLongLS) * 100.0;
}
